Implement basic scrollend event firing

nmoucht opened the pull request:
#### 432af03e413a4e40c73364d8f4711ed344676193
<pre>
Implement basic scrollend event firing
<a href="https://bugs.webkit.org/show_bug.cgi?id=295661">https://bugs.webkit.org/show_bug.cgi?id=295661</a>
<a href="https://rdar.apple.com/155466666">rdar://155466666</a>

Reviewed by Simon Fraser.

Implement basic scrollend event firing by adding scrollend event
target list on Document, and populating it via
AsyncScrollingCoordinator callbacks from the UI process indicating
that a scroll ended for a particular element.

* LayoutTests/fast/scrolling/mac/scrollend-event-user-scroll-basic-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/scrollend-event-user-scroll-basic.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::commonTeardown):
(WebCore::Document::addPendingScrollendEventTarget):
(WebCore::Document::runScrollSteps):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollDidEnd):
(WebCore::LocalFrameView::scheduleScrollEvent):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::wheelEventScrollDidEndForNode):
* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::scrollDidEnd):
(WebCore::ScrollableArea::isAwaitingScrollend const):
(WebCore::ScrollableArea::setIsAwaitingScrollend):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::scrollTo):
(WebCore::RenderLayerScrollableArea::scrollDidEnd):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
(WebKit::RemoteScrollingTree::scrollingTreeNodeDidStopWheelEventScroll):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeDidStopWheelEventScroll):

Canonical link: <a href="https://commits.webkit.org/297567@main">https://commits.webkit.org/297567@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f384fe08bc63f92cf68073ec22c988f9d64a7547

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112186 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31917 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22395 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118263 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/62484 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114148 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32570 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40481 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/85227 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/62484 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115133 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25987 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100938 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65658 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25293 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19074 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62108 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95373 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19151 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121589 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39260 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29203 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/94040 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39641 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97179 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93863 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39093 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16885 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/35288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18081 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39148 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44636 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38783 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42120 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40526 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -5550,6 +5551,20 @@ void Document::flushDeferredResizeEvents()
+        m_pendingScrollendEventTargetList = makeUnique<PendingScrollEventTargetList>();
+
+    auto& targets = m_pendingScrollendEventTargetList->targets;
+    if (targets.findIf([&] (auto& entry) { return entry.ptr() == &target; }) != notFound)
```
Can use `!containsIf()`

nmoucht replied:
Fixed.

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -5629,6 +5644,14 @@ void Document::runScrollSteps()
     }
+    if (m_pendingScrollendEventTargetList && !m_pendingScrollendEventTargetList->targets.isEmpty()) {
+        LOG_WITH_STREAM(Events, stream << "Document " << this << " sending scrollend events to pending scrollend event targets");
+        auto currentTargets = WTFMove(m_pendingScrollendEventTargetList->targets);
```
should use `std::exchange()` instead of `WTFMove()` here as `m_pendingScrollendEventTargetList->targets` may get reused later on.

nmoucht replied:
Fixed.

cdumez commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -439,6 +441,16 @@ void RenderLayerScrollableArea::scrollTo(const ScrollPosition& position)
+    if (!isAwaitingScrollend())
+        return;
+    setIsAwaitingScrollend(false);
+    auto& renderer = m_layer.renderer();
```
nit: Unnecessary local variable.

nmoucht replied:
Fixed.

nmoucht committed:
Implement basic scrollend event firing
https://bugs.webkit.org/show_bug.cgi?id=295661
rdar://155466666

Reviewed by Simon Fraser.

Implement basic scrollend event firing by adding scrollend event
target list on Document, and populating it via
AsyncScrollingCoordinator callbacks from the UI process indicating
that a scroll ended for a particular element.

* LayoutTests/fast/scrolling/mac/scrollend-event-user-scroll-basic-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/scrollend-event-user-scroll-basic.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::commonTeardown):
(WebCore::Document::addPendingScrollendEventTarget):
(WebCore::Document::runScrollSteps):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollDidEnd):
(WebCore::LocalFrameView::scheduleScrollEvent):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::wheelEventScrollDidEndForNode):
* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::scrollDidEnd):
(WebCore::ScrollableArea::isAwaitingScrollend const):
(WebCore::ScrollableArea::setIsAwaitingScrollend):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::scrollTo):
(WebCore::RenderLayerScrollableArea::scrollDidEnd):
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
(WebKit::RemoteScrollingTree::scrollingTreeNodeDidStopWheelEventScroll):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeDidStopWheelEventScroll):

Canonical link: https://commits.webkit.org/297567@main
